File: 2018/PHP/5.php
#!/usr/bin/env php
<?php

function fully_react($polymer) {
    do {
        $polymer = preg_replace('/(.)(?!\1)(?i)(\1)/', '', $polymer, -1, $count);
    } while ($count);
    return $polymer;
}

function reduce($polymer, $unit) {
    return preg_replace("/$unit/i", '', $polymer);
}


$input = trim(stream_get_contents(STDIN));
$reacted_input = fully_react($input);
echo strlen($reacted_input) . PHP_EOL;

$unit_types = array_unique(str_split(strtolower($input)));
$lengths = [];
foreach ($unit_types as $unit) {
    $lengths[] = strlen(fully_react(reduce($reacted_input, $unit)));
}
echo min($lengths) . PHP_EOL;

File: 2018/Perl/5.pl
#!/usr/bin/env perl
use feature qw(say);
use List::Util 'min';

sub fully_react {
    my ($polymer) = @_;
    do {
        $count = $polymer =~ s/(.)(?!\1)(?i)(\1)//g;
    } while ($count);
    return $polymer;
}

sub reduce {
    my ($polymer, $unit) = @_;
    $polymer =~ s/$unit//gi;
    return $polymer;
}


my $input = <>;
$input =~ s/^\s+|\s+$//g;

my $reacted_input = fully_react $input;
say length $reacted_input;

my @unit_types = split //, (lc $input) =~ s/(.)(?=.*?\1)//rg;
my @lengths;
for my $unit (@unit_types) {
    push @lengths, length(fully_react(reduce($reacted_input, $unit)))
}
say min @lengths
